File: handy-python/quiz/q20_linked_list_cycle_easy.py
"""
Q20) Linked List Cycle

AI-BEST:
- Floyd's tortoise-and-hare pointers.
- Time: O(n), Space: O(1)
"""


class ListNode:
    def __init__(self, val: int = 0, next: "ListNode | None" = None) -> None:
        self.val = val
        self.next = next


def has_cycle_best(head: ListNode | None) -> bool:
    # Step 1: Move one pointer slowly and the other quickly.
    # Why: If there is a loop, the fast pointer must eventually lap the slow one.
    slow = head
    fast = head

    # Step 2: Compare pointers while the fast runner can still move.
    # Why: A meeting proves that the list circles back on itself.
    while fast is not None and fast.next is not None:
        slow = slow.next
        fast = fast.next.next
        if slow is fast:
            return True

    # Step 3: Return false when the fast pointer hits the end.
    # Why: Reaching null means there is no cycle to loop back into.
    return False


def build_list(values: list[int], cycle_at: int | None = None) -> ListNode | None:
    if not values:
        return None

    nodes = [ListNode(value) for value in values]
    for i in range(len(nodes) - 1):
        nodes[i].next = nodes[i + 1]

    if cycle_at is not None:
        nodes[-1].next = nodes[cycle_at]

    return nodes[0]


def run_demo() -> None:
    print("Q20: Linked List Cycle")
    cases = [
        ([3, 2, 0, -4], 1),
        ([1, 2], 0),
        ([1], None),
    ]
    for values, cycle_at in cases:
        head = build_list(values, cycle_at)
        print(f"Input: values={values}, cycle_at={cycle_at}")
        print("  BEST:", has_cycle_best(head))


if __name__ == "__main__":
    run_demo()

File: handy-ts/quiz/q20_linked_list_cycle_easy.ts
/**
 * Q20) Linked List Cycle
 *
 * AI-BEST:
 * Floyd's tortoise-and-hare pointers.
 * Time: O(n), Space: O(1)
 */

class ListNode {
  constructor(
    public val: number,
    public next: ListNode | null = null,
  ) {}
}

function hasCycleBest(head: ListNode | null): boolean {
  let slow = head;
  let fast = head;

  // Step 1: Move one pointer slowly and one pointer quickly because equal positions can only happen if a loop exists.
  while (fast !== null && fast.next !== null) {
    slow = slow!.next;
    fast = fast.next.next;

    // Step 2: Compare pointers after each move because meeting inside the list proves the presence of a cycle.
    if (slow === fast) return true;
  }

  // Step 3: Return false when the fast pointer falls off the list because that means there was no cycle to trap it.
  return false;
}

function buildList(values: number[], cycleIndex: number | null = null): ListNode | null {
  if (values.length === 0) return null;

  const nodes = values.map((value) => new ListNode(value));
  for (let i = 0; i < nodes.length - 1; i++) {
    nodes[i].next = nodes[i + 1];
  }
  if (cycleIndex !== null && cycleIndex >= 0 && cycleIndex < nodes.length) {
    nodes[nodes.length - 1].next = nodes[cycleIndex];
  }
  return nodes[0];
}

function runDemo(): void {
  console.log("Q20: Linked List Cycle");
  const cases = [
    { values: [3, 2, 0, -4], cycleIndex: 1 },
    { values: [1, 2, 3], cycleIndex: null },
    { values: [1], cycleIndex: 0 },
  ];

  for (const { values, cycleIndex } of cases) {
    const head = buildList(values, cycleIndex);
    console.log(`Input: values=${JSON.stringify(values)}, cycleIndex=${cycleIndex}`);
    console.log("  BEST:", hasCycleBest(head));
  }
}

runDemo();

export {};
